### PR TITLE
V0.1.7

### PR DESCRIPTION
**Pull request by rglss**
Resolves issues with:
 - Extraction of SectionTypeDesc from units (not always present)
 - Writing cache files when lacking filesystem write permissions
 - Writing to certifi PEM when lacking filesystem write permissions

### CODE REVIEW

File: compass/hierarchy.py [hunk 1]
```diff
@@ -150,9 +152,15 @@ def get_hierarchy(self, compass_id: int, level: str) -> dict:
...
         out = self._get_descendants_recursive(compass_id, hier_level=level)
-        with open(filename, "w", encoding="utf-8") as f:
-            json.dump(out, f, ensure_ascii=False)
+
+        # Try and write to a file for caching
+        try:
+            with open(filename, "w", encoding="utf-8") as f:
+                json.dump(out, f, ensure_ascii=False)
+        except IOError as x:
+            print(f"Unable to write cache file: {x.errno} - {x.strerror}")
```
Comment (AA-Turner) [inline compass/hierarchy.py:163]: I assume these are for convenience (I ususally have write permissions locally)! 

Perhaps add a helper function in utility.py as the try/except code is exactly the same?

```suggestion
        except IOError as e:
            print(f"Unable to write cache file: {e.errno} - {e.strerror}")
```

File: compass/hierarchy.py [hunk 2]
```diff
@@ -243,12 +251,17 @@ def _get_all_members_in_hierarchy(self, parent_id: int, compass_ids: pd.Series)
...
             all_members[compass_id] = self._scraper.get_members_with_roles_in_unit(compass_id)
 
-        with open(f"all-members-{parent_id}.json", "w", encoding="utf-8") as f:
-            json.dump(all_members, f, ensure_ascii=False, indent=4)
+        # Try and write to a file for caching
+        try:
+            with open(f"all-members-{parent_id}.json", "w", encoding="utf-8") as f:
+                json.dump(all_members, f, ensure_ascii=False, indent=4)
+        except IOError as x:
+            print(f"Unable to write cache file: {x.errno} - {x.strerror}")
```
Comment (AA-Turner) [inline compass/hierarchy.py:265]:
```suggestion
        except IOError as e:
            print(f"Unable to write cache file: {e.errno} - {e.strerror}")
```

File: compass/utility.py [hunk 3]
```diff
@@ -30,9 +30,13 @@ def setup_tls_certs() -> None:
...
-        with certifi_path.open("a", encoding="utf-8") as f:
-            f.write('\n# Label: "Thawte RSA CA 2018"\n')
-            f.write(ca_request.text)
+        # Write to certifi PEM
+        try:
+            with certifi_path.open("a", encoding="utf-8") as f:
+                f.write('\n# Label: "Thawte RSA CA 2018"\n')
+                f.write(ca_request.text)
+        except IOError as x:
+            print(f"Unable to write to certifi PEM: {x.errno} - {x.strerror}")
```
Comment (AA-Turner) [inline compass/utility.py:39]:
```suggestion
        except IOError as e:
            print(f"Unable to write to certifi PEM: {e.errno} - {e.strerror}")
```

**Review by AA-Turner — CHANGES_REQUESTED** [review]
Looks good, couple suggestions. Also maybe use pathlib for the JSON stuff but I can do that in my upcoming mega-PR

**Comment by rglss** [comment]
Changes are primarily to handle usage on Azure Functions which has limited write permissions, plus it's always good practice to actually check you can write a file.

Feel it a little overkill to pull this out into a standalone function...

**Pull request merged**